File: src/main/java/in/dragonbra/javasteam/base/AClientMsgProtobuf.java
package in.dragonbra.javasteam.base;

import in.dragonbra.javasteam.enums.EMsg;
import in.dragonbra.javasteam.generated.MsgHdrProtoBuf;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesBase.CMsgProtoBufHeader;
import in.dragonbra.javasteam.types.JobID;
import in.dragonbra.javasteam.types.SteamID;
import in.dragonbra.javasteam.util.log.LogManager;
import in.dragonbra.javasteam.util.log.Logger;

import java.io.ByteArrayInputStream;
import java.io.IOException;

/**
 * Represents a protobuf backed client message. Only contains the header information.
 */
public class AClientMsgProtobuf extends MsgBase<MsgHdrProtoBuf> {

    private static final Logger logger = LogManager.getLogger(AClientMsgProtobuf.class);

    /**
     * Initializes a new instance of the{@link AClientMsgProtobuf} class.
     * This is a recieve constructor.
     *
     * @param msg The packet message to build this client message from.
     */
    public AClientMsgProtobuf(IPacketMsg msg) {
        this(msg.getMsgType());

        if (!msg.isProto()) {
            logger.debug("ClientMsgProtobuf used for non-proto message!");
        }

        deserialize(msg.getData());
    }

    private AClientMsgProtobuf() {
        this(0);
    }

    AClientMsgProtobuf(int payloadReserve) {
        super(MsgHdrProtoBuf.class, payloadReserve);
    }

    private AClientMsgProtobuf(EMsg eMsg) {
        this(eMsg, 0);
    }

    private AClientMsgProtobuf(EMsg eMsg, int payloadReserve) {
        super(MsgHdrProtoBuf.class, payloadReserve);
        // set our emsg
        getHeader().setEMsg(eMsg);
    }

    public CMsgProtoBufHeader.Builder getProtoHeader() {
        return getHeader().getProto();
    }

    @Override
    public boolean isProto() {
        return true;
    }

    @Override
    public EMsg getMsgType() {
        return getHeader().getMsg();
    }

    @Override
    public int getSessionID() {
        return getProtoHeader().getClientSessionid();
    }

    @Override
    public void setSessionID(int sessionID) {
        getProtoHeader().setClientSessionid(sessionID);
    }

    @Override
    public SteamID getSteamID() {
        return new SteamID(getProtoHeader().getSteamid());
    }

    @Override
    public void setSteamID(SteamID steamID) {
        if (steamID == null) {
            throw new IllegalArgumentException("steamID is null");
        }
        getProtoHeader().setSteamid(steamID.convertToUInt64());
    }

    @Override
    public JobID getTargetJobID() {
        return new JobID(getProtoHeader().getJobidTarget());
    }

    @Override
    public void setTargetJobID(JobID jobID) {
        if (jobID == null) {
            throw new IllegalArgumentException("jobID is null");
        }
        getProtoHeader().setJobidTarget(jobID.getValue());
    }

    @Override
    public JobID getSourceJobID() {
        return new JobID(getProtoHeader().getJobidSource());
    }

    @Override
    public void setSourceJobID(JobID jobID) {
        if (jobID == null) {
            throw new IllegalArgumentException("jobID is null");
        }
        getProtoHeader().setJobidSource(jobID.getValue());
    }

    @Override
    public byte[] serialize() {
        throw new UnsupportedOperationException("ClientMsgProtobuf is for reading only. Use ClientMsgProtobuf<T> for serializing messages.");
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            getHeader().deserialize(new ByteArrayInputStream(data));
        } catch (IOException e) {
            logger.debug(e);
        }
    }
}

File: src/main/java/in/dragonbra/javasteam/base/AClientMsgProtobuf.kt
package `in`.dragonbra.javasteam.base

import `in`.dragonbra.javasteam.enums.EMsg
import `in`.dragonbra.javasteam.generated.MsgHdrProtoBuf
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesBase
import `in`.dragonbra.javasteam.types.JobID
import `in`.dragonbra.javasteam.types.SteamID
import `in`.dragonbra.javasteam.util.log.LogManager
import `in`.dragonbra.javasteam.util.log.Logger
import `in`.dragonbra.javasteam.util.stream.MemoryStream
import java.io.IOException

/**
 * Represents a protobuf backed client message. Only contains the header information.
 */
@Suppress("unused")
open class AClientMsgProtobuf : MsgBase<MsgHdrProtoBuf> {

    companion object {
        private val logger: Logger = LogManager.getLogger(AClientMsgProtobuf::class.java)
    }

    /**
     * Shorthand accessor for the protobuf header.
     */
    val protoHeader: SteammessagesBase.CMsgProtoBufHeader.Builder
        get() = header.proto

    /**
     * Gets a value indicating whether this client message is protobuf backed.
     * Client messages of this type are always protobuf backed.
     * @return **true** if this instance is protobuf backed; otherwise, **false**.
     */
    override val isProto: Boolean
        get() = true

    /**
     * Gets the network message type of this client message.
     * @return The network message type.
     */
    override val msgType: EMsg
        get() = header.msg

    /**
     * Gets or Sets the session id for this client message.
     */
    override var sessionID: Int
        get() = protoHeader.clientSessionid
        set(value) {
            protoHeader.clientSessionid = value
        }

    /**
     * Gets or Sets the [SteamID] for this client message.
     */
    override var steamID: SteamID?
        get() = SteamID(protoHeader.steamid)
        set(value) {
            protoHeader.steamid = value!!.convertToUInt64()
        }

    /**
     * Gets or Sets the target job id for this client message.
     */
    override var targetJobID: JobID
        get() = JobID(protoHeader.jobidTarget)
        set(value) {
            protoHeader.jobidTarget = value.value
        }

    /**
     * Gets or Sets the source job id for this client message.
     */
    override var sourceJobID: JobID
        get() = JobID(protoHeader.jobidSource)
        set(value) {
            protoHeader.jobidSource = value.value
        }

    /**
     * Initializes a new instance of the [AClientMsgProtobuf] class.
     * This is a recieve constructor.
     * @param msg The packet message to build this client message from.
     */
    constructor(msg: IPacketMsg) : this(msg.msgType) {
        if (!msg.isProto) {
            logger.error("ClientMsgProtobuf used for non-proto message!")
        }

        deserialize(msg.data)
    }

    private constructor() : this(0)

    internal constructor(payloadReserve: Int) : super(MsgHdrProtoBuf::class.java, payloadReserve)

    private constructor(eMsg: EMsg, payloadReserve: Int = 0) : super(MsgHdrProtoBuf::class.java, payloadReserve) {
        // set our emsg
        header.setEMsg(eMsg)
    }

    /**
     * Serializes this client message instance to a byte array.
     * @throws UnsupportedOperationException This class is for reading Protobuf messages only. If you want to create a protobuf message, use [ClientMsgProtobuf].
     */
    override fun serialize(): ByteArray = throw UnsupportedOperationException("ClientMsgProtobuf is for reading only. Use ClientMsgProtobuf<T> for serializing messages.")

    /**
     * Initializes this client message by deserializing the specified data.
     * @param data The data representing a client message.
     */
    override fun deserialize(data: ByteArray) {
        try {
            MemoryStream(data).use { ms ->
                header.deserialize(ms)
            }
        } catch (e: IOException) {
            logger.error(e)
        }
    }
}
